### ISSUE TITLE
Memory leak at nla_intervals.cpp

### ISSUE THREAD

**Issue by rainoftime** [Z3Prover/z3#3638]
Hi, for the following formula,
~~~~
(set-logic NRA)
(set-option :rewriter.flat false)
(set-option :smt.arith.solver 6)
(set-option :smt.random_seed 5)
(declare-const v5 Bool)
(declare-const r1 Real)
(declare-const v17 Bool)
(declare-const r2 Real)
(push 1)
(assert (or v17 (> r2 (- r1) (/ r1 0.0) (* r1 (- r1) r1 20.0) 0.0) v5))
(check-sat
~~~~

asan detects a memory leak issue in z3 (commit 1d6fb63)
~~~~
irect leak of 120 byte(s) in 3 object(s) allocated from:
    #0 0x7fd965b73662 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98662)
    #1 0x255cabb in memory::allocate(unsigned long) ../src/util/memory_manager.cpp:268
    #2 0x258ca8b in small_object_allocator::allocate(unsigned long) ../src/util/small_object_allocator.cpp:103
    #3 0x25a4c27 in mpz_manager<false>::allocate(unsigned int) ../src/util/mpz.cpp:198
    #4 0x25a738f in mpz_manager<false>::big_set(mpz&, mpz const&) ../src/util/mpz.cpp:1497
    #5 0x562489 in mpz_manager<false>::set(mpz&, mpz const&) ../src/util/mpz.h:530
    #6 0x563ebe in mpq_manager<false>::set(mpz&, mpz const&) ../src/util/mpq.h:653
    #7 0x563a70 in mpq_manager<false>::set(mpq&, mpq const&) ../src/util/mpq.h:656
    #8 0x1e25880 in interval_manager<dep_intervals::im_config>::set(dep_intervals::im_config::interval&, dep_intervals::im_config::interval const&) ../src/math/interval/interval_def.h:680
    #9 0x1e2bd53 in void dep_intervals::set<(dep_intervals::with_deps_t)1>(dep_intervals::im_config::interval&, dep_intervals::im_config::interval const&) const (/home/test/tofuzz/z3-debug/build/z3
+0x1e2bd53)
    #10 0x1e27707 in dep_intervals::im_config::interval nla::intervals::interval_of_mul<(dep_intervals::with_deps_t)1>(nla::nex_mul const&) ../src/math/lp/nla_intervals.cpp:383
    #11 0x1e243fb in dep_intervals::im_config::interval nla::intervals::interval_of_expr<(dep_intervals::with_deps_t)1>(nla::nex const*, unsigned int) ../src/math/lp/nla_intervals.cpp:408
    #12 0x1e2ae99 in dep_intervals::im_config::interval nla::intervals::interval_of_sum_no_term<(dep_intervals::with_deps_t)1>(nla::nex_sum const&) (/home/test/tofuzz/z3-debug/build/z3+0x1e2ae99)
    #13 0x1e26a79 in dep_intervals::im_config::interval nla::intervals::interval_of_sum<(dep_intervals::with_deps_t)1>(nla::nex_sum const&) ../src/math/lp/nla_intervals.cpp:337
    #14 0x1e24382 in dep_intervals::im_config::interval nla::intervals::interval_of_expr<(dep_intervals::with_deps_t)1>(nla::nex const*, unsigned int) ../src/math/lp/nla_intervals.cpp:402
    #15 0x1e1dd5a in nla::intervals::check_nex(nla::nex const*, dependency_manager<scoped_dependency_manager<unsigned int>::config>::dependency*) ../src/math/lp/nla_intervals.cpp:75
    #16 0x1dbae48 in bool nla::horner::lemmas_on_row<old_vector<lp::row_cell<rational>, true, unsigned int> >(old_vector<lp::row_cell<rational>, true, unsigned int> const&)::{lambda(nla::nex const*)#1}::o
perator()(nla::nex const*) const (/home/test/tofuzz/z3-debug/build/z3+0x1dbae48)
    #17 0x1dbf690 in std::_Function_handler<bool (nla::nex const*), bool nla::horner::lemmas_on_row<old_vector<lp::row_cell<rational>, true, unsigned int> >(old_vector<lp::row_cell<rational>, true, unsign
ed int> const&)::{lambda(nla::nex const*)#1}>::_M_invoke(std::_Any_data const&, nla::nex const*&&) /usr/include/c++/5/functional:1857
    #18 0x1dba4ae in std::function<bool (nla::nex const*)>::operator()(nla::nex const*) const (/home/test/tofuzz/z3-debug/build/z3+0x1dba4ae)
    #19 0x1db56d6 in nla::cross_nested::explore_of_expr_on_sum_and_var(nla::nex**, unsigned int, old_vector<nla::nex**, true, unsigned int>) ../src/math/lp/cross_nested.h:287
    #20 0x1db38a4 in nla::cross_nested::explore_expr_on_front_elem_vars(nla::nex**, old_vector<nla::nex**, true, unsigned int>&, old_svector<unsigned int, unsigned int> const&) ../src/math/lp/cross_nested
.h:173
    #21 0x1db4f74 in nla::cross_nested::explore_expr_on_front_elem(nla::nex**, old_vector<nla::nex**, true, unsigned int>&) ../src/math/lp/cross_nested.h:266
    #22 0x1db1d8f in nla::cross_nested::run(nla::nex*) ../src/math/lp/cross_nested.h:70
    #23 0x1dae1e0 in nla::horner::lemmas_on_expr(nla::cross_nested&, nla::nex_sum*) ../src/math/lp/horner.cpp:68
    #24 0x1dbb374 in bool nla::horner::lemmas_on_row<old_vector<lp::row_cell<rational>, true, unsigned int> >(old_vector<lp::row_cell<rational>, true, unsigned int> const&) ../src/math/lp/horner.cpp:90
    #25 0x1dae88c in nla::horner::horner_lemmas() ../src/math/lp/horner.cpp:121
    #26 0x1d864bc in nla::core::inner_check(bool) ../src/math/lp/nla_core.cpp:1291
    #27 0x1d87d3e in nla::core::check(old_vector<nla::lemma, true, unsigned int>&) ../src/math/lp/nla_core.cpp:1422
    #28 0x1c817c3 in nla::solver::check(old_vector<nla::lemma, true, unsigned int>&) ../src/math/lp/nla_solver.cpp:40
    #29 0x119892b in smt::theory_lra::imp::check_nra() ../src/smt/theory_lra.cpp:2189
~~~~

**Comment by NikolajBjorner** [comment]
@levnach many functions in nla_intervals need to be rewritten to use scoped_dep_interval instead of dep_intervals::interval. The big-nums that are contained in the interval structure are not deleted automatically. They are owned by the dep_interval class. 
The class pdd_interval.h illustrates one use of scoped intervals.

Functions, such as 

        interval interval_of_sum(const nex_sum& e);

could instead have signature

        void interval_of_sum(const nex_sum& e, scoped_dep_interval& result);

**Comment by levnach** [comment]
Should be fixed by afce09efe4848f35d84106200db4859dbbb550fd
